mlkem/ntt.c

```diff
@@ -180,8 +180,11 @@ void basemul(int16_t r[2], const int16_t a[2], const int16_t b[2],
  **************************************************/
 void basemul_cached(int16_t r[2], const int16_t a[2], const int16_t b[2], int16_t b_cached)
 {
-    r[0] = fqmul(a[1], b_cached);
-    r[0] += fqmul(a[0], b[0]);
-    r[1] = fqmul(a[0], b[1]);
-    r[1] += fqmul(a[1], b[0]);
+    int32_t t0, t1;
+    t0  = (int32_t) a[1] * b_cached;
+    t0 += (int32_t) a[0] * b[0];
+    t1  = (int32_t) a[0] * b[1];
+    t1 += (int32_t) a[1] * b[0];
+    r[0] = montgomery_reduce(t0);
```

Comment:
Do t0 and t1 meet the precondition on montgomery_reduce()?

Reply:
I think we have an issue #43 to document bounds everywhere (and ultimately prove them in CBMC). I _think_ we can assume here that `a,b` are signed canonical.

Reply:
In my code, I have
```
#define MONT_DOMAIN_LB (-32768 * KYBER_Q)
#define MONT_DOMAIN_UB (32768 * KYBER_Q - 1)

static inline int16_t montgomery_reduce(int32_t a)
__CPROVER_requires(a >= MONT_DOMAIN_LB)
__CPROVER_requires(a <= MONT_DOMAIN_UB)
```
Can you justify that's OK here?

Reply:
If `a,b` are bounded by `Q` in absolute value, then any `a[i] * b[j]` is bound by `Q^2 < 2^12 Q`. This means that `t` is bound by `4*2^12*Q < 2^15 * Q`.